### PR TITLE
feat(snuba): Inject meta into snuba results converter [TET-131]

### PR DESCRIPTION
**Pull request by andriisoldatenko**
Add meta data in the resulting response from get_series
as we are just discarding it and returning the coalesced data portion.

### CODE REVIEW

File: src/sentry/snuba/metrics/query_builder.py [hunk 1]
```diff
@@ -642,28 +656,37 @@ def translate_results(self):
...
-                    )
+                    public_metric_key = self._get_operation_with_public_name(operation, metric_mri)
                     if totals is not None:
                         totals[public_metric_key] = totals.pop(key)
                     if series is not None:
                         series[public_metric_key] = series.pop(key)
 
         return groups
+
+    def merge_result_meta(self):
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query_builder.py:674]: remove from class.

File: src/sentry/snuba/metrics/query_builder.py [hunk 2]
```diff
@@ -587,9 +589,21 @@ def _extract_data(self, data, groups):
...
                             series[series_index] = cleaned_value
 
-    def translate_results(self):
+    def _get_operation_with_public_name(self, operation: Optional[str], metric_mri: str) -> str:
+        if operation is None:
+            return get_public_name_from_mri(metric_mri)
+        return f"{operation}({get_public_name_from_mri(metric_mri)})"
+
+    def _parse_operation_and_metric_mri(self, name: str) -> Tuple[Optional[str], str]:
+        matches = MRI_EXPRESSION_REGEX.match(name)
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query_builder.py:598]: replace with MRI_NAME_REGEX

File: src/sentry/snuba/metrics/query_builder.py [hunk 3]
```diff
@@ -587,9 +589,21 @@ def _extract_data(self, data, groups):
...
+        if matches:
+            # operation, metric_mri
+            return matches[1], matches[2]
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query_builder.py:601]: `if MRI_NAME_REGEX.matches....`

Comment (andriisoldatenko) [reply src/sentry/snuba/metrics/query_builder.py]: after deep dive: we can't use `MRI_NAME_REGEX = r"([a-z_]+(?:\\.[a-z_]+)*)"` here it breaks things https://github.com/getsentry/sentry/blob/9441adfc40a7923bbbf1c663988ad15c26511abb/src/sentry/snuba/metrics/query_builder.py#L683

This function should only accept metric with operation or fall back to metric without operation.

File: tests/sentry/snuba/metrics/test_query_builder.py [hunk 4]
```diff
@@ -844,3 +845,32 @@ def test_translate_results_missing_slots(_1, _2, monkeypatch):
...
+        {"name": "p50(d:transactions/measurements.lcp@millisecond)", "type": "Array(Float64)"},
+        {"name": "p50(d:transactions/measurements.lcp@millisecond)", "type": "Array(Float64)"},
+        {"name": "tags[9223372036854776020]", "type": "UInt64"},
+        {"name": "tags[9223372036854776020]", "type": "UInt64"},
+        {"name": "project_id", "type": "UInt64"},
+        {"name": "project_id", "type": "UInt64"},
+    ]
+    assert translate_meta_results(meta) == [
+        {"name": "p50(transaction.measurements.lcp)", "type": "Array(Float64)"},
+        {"name": "transaction", "type": "UInt64"},
```
Comment (andriisoldatenko) [inline tests/sentry/snuba/metrics/test_query_builder.py:874]: @ahmedetefy i know we dicussed it, but i can't find correct answer in code. I need valid name of `String` type here.

File: tests/sentry/snuba/metrics/test_query_builder.py [hunk 5]
```diff
@@ -844,3 +845,32 @@ def test_translate_results_missing_slots(_1, _2, monkeypatch):
...
+        {"name": "project_id", "type": "UInt64"},
+    ]
+    assert translate_meta_results(meta) == [
+        {"name": "p50(transaction.measurements.lcp)", "type": "Array(Float64)"},
+        {"name": "transaction", "type": "string"},
+        {"name": "project_id", "type": "UInt64"},
+    ]
+
+
+def test_translate_meta_results_with_duplicates():
```
Comment (ahmedetefy) [inline tests/sentry/snuba/metrics/test_query_builder.py:863]: Great unit tests! Thanks for that

File: src/sentry/snuba/metrics/datasource.py [hunk 6]
```diff
@@ -678,36 +688,46 @@ def get_series(projects: Sequence[Project], metrics_query: MetricsQuery) -> dict
...
+    # Translate applies only on ["data"]
+    result_groups = converter.translate_result_groups()
     # It can occur, when we make queries that are not ordered, that we end up with a number of
     # groups that doesn't meet the limit of the query for each of the entities, and hence they
     # don't go through the pruning logic resulting in a total number of groups that is greater
     # than the limit, and hence we need to prune those excess groups
     if len(result_groups) > metrics_query.limit.limit:
         result_groups = result_groups[0 : metrics_query.limit.limit]
 
+    metrics_query_fields = {str(metric_field) for metric_field in metrics_query.select}
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/datasource.py:725]:
```suggestion
    metrics_query_fields = set(map(str, metrics_query.select))
```

In [3]: set(map(str, a))
Out[3]: {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'}

Comment (ahmedetefy) [reply src/sentry/snuba/metrics/datasource.py:725]: I prefer the set comprehension to be honest

Comment (andriisoldatenko) [reply src/sentry/snuba/metrics/datasource.py:725]: sure

**Review by ahmedetefy — APPROVED** [review]

**Pull request merged**